lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -1018,7 +1068,13 @@ bool DxilLoopUnroll::runOnLoop(Loop *L, LPPassManager &LPM) {
       for (size_t i = 0; i < Iterations.size(); i++) {
         LoopIteration &Iteration = *Iterations[i].get();
         for (BasicBlock *BB : Iteration.Body) {
-          if (!Iteration.Extended.count(BB)) {
+          if (!Iteration.Extended.count(BB) &&
+            // Check We didn't already add BB to a sub-loop
+            !LI->getLoopFor(BB) &&
+            // Check if it's already added to loop when we added
+            // the block to new child loop earlier.
+            !OuterL->contains(BB))
```

Comment:
Is there a correct case where it has been added to a subloop, but that subloop isn't part of the outer loop?

What I'm asking is whether it's possible for `LI->getLoopFor(BB)  && !OuterL->contains(BB)` to be true. If not, I think you can just check the `OuterL->contains(BB)` and maybe assert the other part?

Reply:
In an older version of this change, RecursivelyRecreateSubLoopForIteration didn't add all blocks to all outer loops. That's no longer the case.

Reply:
Fixed.